src/libcephfs_proxy/proxy_link.c

```diff
@@ -55,6 +68,523 @@ static int32_t proxy_link_prepare(struct sockaddr_un *addr, const char *path)
+
+	cmsg = (struct cmsghdr *)buffer;
+	cmsg->cmsg_len = CMSG_LEN(ctrl_size);
+	cmsg->cmsg_level = SOL_SOCKET;
+	cmsg->cmsg_type = type;
+	memcpy(CMSG_DATA(cmsg), ctrl, ctrl_size);
+	msg.msg_control = buffer;
+	msg.msg_controllen = sizeof(buffer);
+
+	msg.msg_flags = 0;
+
+	len = sendmsg(sd, &msg, MSG_NOSIGNAL);
+	if (len < 0) {
+		return proxy_log(LOG_ERR, errno,
+				 "Failed to send a control message");
+	}
+
+	if (len != size) {
+		return proxy_log(LOG_ERR, 0,
+				 "Control message not sent completely");
+	}
+
+	return 0;
+}
+
+int32_t proxy_link_ctrl_recv(int32_t sd, void *data, int32_t size, int32_t type,
+			     void *ctrl, int32_t *ctrl_size)
+{
+	char buffer[CMSG_SPACE(*ctrl_size)];
+	struct msghdr msg;
+	struct iovec iov;
+	struct cmsghdr *cmsg;
+	ssize_t len;
+	int32_t clen;
+
+	msg.msg_name = NULL;
+	msg.msg_namelen = 0;
+
+	iov.iov_base = data;
+	iov.iov_len = size;
+	msg.msg_iov = &iov;
+	msg.msg_iovlen = 1;
+
+	msg.msg_control = buffer;
+	msg.msg_controllen = sizeof(buffer);
+
+	msg.msg_flags = 0;
+
+	len = recvmsg(sd, &msg, MSG_NOSIGNAL | MSG_CMSG_CLOEXEC);
+	if (len < 0) {
+		return proxy_log(LOG_ERR, errno,
+				 "Failed to received a control message");
+	}
+
+	if (len != size) {
+		return proxy_log(LOG_ERR, 0,
+				 "Message data not received completely");
+	}
+
+	if ((msg.msg_flags & MSG_CTRUNC) != 0) {
+		return proxy_log(LOG_ERR, ENOSPC,
+				 "Message data is larger than expected");
+	}
+
+	cmsg = CMSG_FIRSTHDR(&msg);
+	if (cmsg == NULL) {
+		*ctrl_size = 0;
+		return len;
+	}
+
+	if (cmsg->cmsg_level != SOL_SOCKET) {
+		return proxy_log(LOG_ERR, EINVAL,
+				 "Unexpected control message level");
+	}
+
+	if (cmsg->cmsg_type != type) {
+		return proxy_log(LOG_ERR, EINVAL,
+				 "Unexpected control message type");
+	}
+
+	clen = cmsg->cmsg_len - CMSG_LEN(0);
+	if (clen != *ctrl_size) {
+		return proxy_log(LOG_ERR, EINVAL,
+				 "Unexpected control message size");
+	}
+
+	if (CMSG_NXTHDR(&msg, cmsg) != NULL) {
+		return proxy_log(LOG_ERR, EINVAL, "Too many control messages");
+	}
+
+	memcpy(ctrl, CMSG_DATA(cmsg), clen);
+
+	return len;
+}
+
+static int32_t proxy_link_negotiate_read(proxy_link_t *link, int32_t sd,
+					 proxy_link_negotiate_t *neg)
+{
+	char buffer[128];
+	void *ptr;
+	uint32_t size, len;
+	int32_t err;
+
+	memset(neg, 0, sizeof(proxy_link_negotiate_t));
+
+	ptr = neg;
+
+	err = proxy_link_read(link, sd, ptr, sizeof(neg->v0));
+	if (err < 0) {
+		return err;
+	}
+
+	ptr += sizeof(neg->v0);
+	size = neg->v0.size;
+
+	/* Version 0 is only used to represent legacy peers that don't use any
+	 * negotiation requests. */
+	if (neg->v0.version == 0) {
+		return proxy_log(LOG_ERR, EINVAL, "Invalid negotiate version");
+	}
+
+	/* Ignore data from future versions. */
+	if (neg->v0.version > PROXY_LINK_NEGOTIATE_VERSION) {
+		neg->v0.version = PROXY_LINK_NEGOTIATE_VERSION;
+		neg->v0.size = NEG_VERSION_SIZE(PROXY_LINK_NEGOTIATE_VERSION);
+	}
+
+	if (negotiation_sizes[neg->v0.version] != neg->v0.size) {
```

Comment:
Considering multiple threads, should we hv `rwlock` for `negotiation_sizes`  global array? (thinking of race condition here)

Reply:
`negotiation_sizes` is a read only array. There's no potential race here.